docs/topics/items.rst

```diff
@@ -8,29 +8,124 @@ Items
+<supporting-item-types>`.
+
+.. _item-types:
+
+Item Types
+==========
+
+Scrapy supports the following types of items, via the `itemadapter`_ library:
+:ref:`dictionaries <dict-items>`, :ref:`Item objects <item-objects>`,
+:ref:`dataclass objects <dataclass-items>`, and :ref:`attrs objects <attrs-items>`.
+
+.. _itemadapter: https://github.com/scrapy/itemadapter
+
+.. _dict-items:
+
+Dictionaries
+------------
+
+As an item type, :class:`dict` is convenient and familiar.
+
+.. _item-objects:
+
+Item objects
+------------
+
+:class:`Item` provides a :class:`dict`-like API plus additional features that
+make it the most feature-complete item type:
+
+.. class:: Item([arg])
+
+    :class:`Item` objects replicate the standard :class:`dict` API, including
+    its ``__init__`` method.
+
+    :class:`Item` allows defining field names, so that:
+
+    -   :class:`KeyError` is raised when using undefined field names (i.e.
+        prevents typos going unnoticed)
+
+    -   :ref:`Item exporters <topics-exporters>` can export all fields by
+        default even if the first scraped object does not have values for all
+        of them
+
+    :class:`Item` also allows defining field metadata, which can be used to
+    :ref:`customize serialization <topics-exporters-field-serialization>`.
+
+    :mod:`trackref` tracks :class:`Item` objects to help find memory leaks
+    (see :ref:`topics-leaks-trackrefs`).
+
+    :class:`Item` objects also provide the following additional API members:
+
+    .. automethod:: copy
+
+    .. automethod:: deepcopy
+
+    .. attribute:: fields
+
+        A dictionary containing *all declared fields* for this Item, not only
+        those populated. The keys are the field names and the values are the
+        :class:`Field` objects used in the :ref:`Item declaration
+        <topics-items-declaring>`.
+
+.. _dataclass-items:
+
+Dataclass objects
+-----------------
+
+.. versionadded:: 2.2
+
+:func:`~dataclasses.dataclass` allows defining item classes with field names,
+so that :ref:`item exporters <topics-exporters>` can export all fields by
+default even if the first scraped object does not have values for all of them.
+
+Dataclasses also allow defining the type and default value of each defined
+field.
+
+.. note:: Field names and types are not enforced at run time.
+
+They work natively in Python 3.7 or later, or using the `dataclasses
+backport`_ in Python 3.6.
+
+.. _dataclasses backport: https://pypi.org/project/dataclasses/
+
+.. _attrs-items:
+
+Attr.s objects
+--------------
+
+.. versionadded:: 2.2
+
+:func:`attr.s` allows defining item classes with field names,
+so that :ref:`item exporters <topics-exporters>` can export all fields by
+default even if the first scraped object does not have values for all of them.
+
+`attr.s` also allow defining the type and default value of each defined field.
+
+.. note:: Field names and types are not enforced at run time.
```

Comment:
I _think_ you can have it by defining a class with `@attr.s(slots=True)`

Reply:
I don't really know, but I can test it and see what happens. Let's summon @Gallaecio, who authored this particular change, in case he can provide some insight.

Reply:
For the record, it doesn't seem like `slots=True` enforces types. Also, I couldn't find relevant information at https://www.attrs.org/en/stable/examples.html#types.

```python
In [1]: import attr 
   ...: @attr.s(slots=True) 
   ...: class AttrsItem: 
   ...:     name = attr.ib(type=str, metadata={"serializer": str}) 
   ...:     value = attr.ib(type=int, metadata={"serializer": int}) 
   ...:

In [2]: AttrsItem(name=999, value="not-an-int")
Out[2]: AttrsItem(name=999, value='not-an-int')
```

Reply:
Well, at least they enforce the names, as `Item` does.

What about removing the note from there and refactoring the first list item above as follows?

```
* Define the type and default value of each defined field.

  Use :term:`slotted classes` so that an :exc:`AttributeError` exception
  is raised when using undefined field names (i.e. prevent typos 
  going unnoticed).

  .. note:: Field types are not enforced at run time.

* Define custom field `metadata`_, which can be used to
  :ref:`customize serialization <topics-exporters-field-serialization>`.
```

Notes:
- This requires adding `attr.s` to the InterSphinx mapping.
- I’m not sure if `` :term:`slotted classes` `` will work right away. It may require to prefix the term with the `attr.s` InterSphinx alias, e.g. `` :term:`attrs:slotted classes` ``
- Since we are adding `attr.s` to the InterSphinx mapping, we should use it for the metadata link as well, maybe also the main link (point to https://www.attrs.org/en/stable/index.html, `` :doc:`attrs:index` ``).

Reply:
I changed the note to "Field types are not enforced at run time". I actually think this note might not be necessary in general, because it's not a Scrapy characteristic. I left it for `dataclasses` but removed the `attrs` one, which doesn't even mention types.
I don't think we need to mention slotted classes, that's not specific to Scrapy but to `attrs`. The links to the upstream docs should be enough IMHO.